add : 홈화면과 관련된 모델들 추가 & 테스트 관련 모델들 추가

inwoodev opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경
홈화면과 관련된 모델들을 구현해야 합니다. 특히 게시물 작성 feature와 연결되는 모델들이 많기에 먼저 빠르게 머지가 필요합니다.

## 목표
- 홈화면과 관련된 dto & model 구현
- 필요한 모델 수정

inwoodev committed:
add : 홈화면과 관련된 모델들 추가 & 테스트 관련 모델들 추가

inwoodev commented on Projects/Domain/Shared/Interface/Sources/Models/Lyrics.swift:
```diff
@@ -0,0 +1,88 @@
+        case .beige:            return FeelinImages.image06Beige
+        case .pinkgreen:        return FeelinImages.image07Pinkgreen
+        case .red:              return FeelinImages.image08Red
+        case .white:            return FeelinImages.image09White
+        case .rainbow:          return FeelinImages.image10Rainbow
+        }
+    }
+    
+    public init(dto: LyricsBackgroundDTO) {
+        switch dto {
+        case .default:          self = .default
+        case .skyblue:          self = .skyblue
+        case .blue:             self = .blue
+        case .lavender:         self = .lavender
+        case .mint:             self = .mint
+        case .black:            self = .black
+        case .beige:            self = .beige
+        case .pinkgreen:        self = .pinkgreen
+        case .red:              self = .red
+        case .white:            self = .white
+        case .rainbow:          self = .rainbow
+        }
+    }
+    
+    var toDTO: LyricsBackgroundDTO {
+        switch self {
+        case .default:          return .default
+        case .skyblue:          return .skyblue
+        case .blue:             return .blue
+        case .lavender:         return .lavender
+        case .mint:             return .mint
+        case .black:            return .black
+        case .beige:            return .beige
+        case .pinkgreen:        return .pinkgreen
+        case .red:              return .red
+        case .white:            return .white
+        case .rainbow:          return .rainbow
+        }
+    }
+}
```
가사 백그라운드 이미지와 관련된 모델은 여기 있습니다~

derrickkim0109 commented on Projects/Core/Network/Interface/Sources/DTO/ArtistDTO.swift:
```diff
@@ -1,13 +1,13 @@
 import Foundation
 
-public struct ArtistResponse: Decodable {
+public struct ArtistDTO: Decodable {
```
**P3**

Response에서 DTO로 이름 변경하신 이유가 있을까요?
다른 Response DTO와 네이밍이 통일되면 좋을 것 같은데, 어떻게 생각하세요?

inwoodev replied:
요건 제 실수입니다. 클라이언트에서 서버로 보낼 수 있는 DTO들 뒤에 Response 네이밍을 붙이는건 적절치 않아보여서 몇 몇 을 바꾸었습니다. 예를들면`NoteStatus` 같은 경우 게시물작성시 서버로 post할 때 사용되기에 `NoteStatusResponse` 보다는 `NoteStatusDTO`가 더 적절해 보였습니다.

다만 Artist같은 경우 저희가 직접 해당 모델을 통해 서버로 전송할 일이 없다고 생각이 들기 때문에
ArtistReponse로 네이밍을 수정하는게 더 적절 해 보입니다.

뿐만 아니라 SongDTO 또한 SongResponse로 수정하는게 이 관점에서는 더 적절 해 보이며
UserDTO같은 경우 Codable을 채택하게 하는게 더 적절해 보이네요.

요런 맥락에서 네이밍 어떤가요?

derrickkim0109 replied:
이해했습니다. 
좋습니다. 그렇게 진행하시죠.

inwoodev replied:
현재 DTO로 사용되는 모델들이 추후 Response와 Request가 달라질 경우

편하게 Response와 Request 모델로 분리하셔도 좋을 듯 합니다.

ex): UserDTO -> UserResponse & UserRequest

derrickkim0109 replied:
네 좋습니다~!

derrickkim0109 commented on Projects/Domain/Note/Interface/Sources/Models/Note.swift:
```diff
@@ -0,0 +1,87 @@
+import Foundation
+
+public struct Note: Hashable {
+    private let uuid: UUID = .init()
+    public let id: Int
+    public let content: String
+    public let status: NoteStatus
+    public let createdAt: Date
+    private (set) public var lyrics: Lyrics?
+    public let publisher: User
+    public let song: Song
+    public let commentsCount: Int
+    public let likesCount: Int
+    public let isLiked: Bool
+    public let isBookmarked: Bool
```
**P3**

다른 프로퍼티도 let으로 선언해서 set이 불가능하지 않나요? lyrics만 private (set)으로 선언하신 이유가 궁금합니다.

inwoodev replied:
dto를 주입받아 초기화 시 `else { self.lyrics = nil }` 구문을 빼고 싶어서 위와 같이 구현하였는데요. 통일감 있게 let으로 가시죠~ 수정할게요
``` swift
public init(dto: NoteResponse) {
        self.id = dto.id
        self.content = dto.content
        self.status = NoteStatus(dto: dto.status)
        self.createdAt = dto.createdAt
        if let dtoLyrics = dto.lyrics {
            self.lyrics = Lyrics(dto: dtoLyrics)
        } else {
            self.lyrics = nil
        }
        self.publisher = User(dto: dto.publisher)
        self.song = Song(dto: dto.song)
        self.commentsCount = dto.commentsCount
        self.likesCount = dto.likesCount
        self.isLiked = dto.isLiked
        self.isBookmarked = dto.isBookmarked
    }
```

derrickkim0109 replied:
넵 알겠습니다~!

derrickkim0109 commented on Projects/Domain/Note/Interface/Sources/Models/Note.swift:
```diff
@@ -0,0 +1,87 @@
+    }
+}
+
+public enum NoteStatus: Hashable {
+    case draft
+    case published
```
NoteStatus는 저도 사용하게 될 것 같은데 혹시 DomainShared로 옮겨 주실 수 있을까요?

derrickkim0109 commented on Projects/Domain/Note/Interface/Sources/Models/Note.swift:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+    
+    public var toDTO: NoteStatusDTO {
+        switch self {
+        case .draft:            return .draft
+        case .published:        return .published
+        }
+    }
```
**P3**

toDTO를 프로퍼티로 선언하신 이유가 있을까요?

inwoodev replied:
게시물 post시 간편하게 활용하실 수 있게 간편하게 만들어 봤습니다. 혹시 어떤 부분이 마음에 걸리실까요?

derrickkim0109 replied:
toDTO를 메서드로 리턴해서 사용하지 않고 프로퍼티로 사용하신 이유가 있을지 궁금했습니다.

inwoodev replied:
아하...음 사실 항상 이 원칙을 지키지는 않는데요...외부에서 객체를 주입받지 않으며 연산프로퍼티를 사용했을 때 문제 없다 판단이 들 경우 그냥 위와 같이 연산 프로퍼티를 사용하는 편입니다!!

derrickkim0109 replied:
넵 알겠습니다~!

derrickkim0109 reviewed (commented):
수고하셨습니다. 인우님 

코멘트에 남겼는데, NoteStatus만 DomainShared로 옮겨 주시면 감사하겠습니다.

derrickkim0109 reviewed (approved):
수고하셨습니다~!

inwoodev committed:
refactor : 피드백 반영 리팩토링 진행

- 모델 네이밍 수정
  - Decodable한 객체들을 Response로 네이밍 수정
  - Codable한 객체들을 DTO로 네이밍 수정

모델 이동
- NoteStatus를 DomainShared로 이동

Tuist Dependency 수정
- SharedDomainInterface의 dependency에 Core추가

그외:
- 변경된 모델들 및 import 때문에 빌드가 안되는 문제 수정

derrickkim0109 reviewed (commented):
수고하셨습니다~!
